drivers/adc/adc_max11102_17.c

```diff
@@ -0,0 +1,454 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER 1
+#include "adc_context.h"
+
+LOG_MODULE_REGISTER(max11102_17, CONFIG_ADC_LOG_LEVEL);
+
+struct max11102_17_config {
+	struct spi_dt_spec bus;
+	const struct gpio_dt_spec gpio_chsel;
+	uint8_t resolution;
+	uint8_t channel_count;
+};
+
+struct max11102_17_data {
+	struct adc_context ctx;
+	struct k_sem acquire_signal;
+	int16_t *buffer;
+	int16_t *buffer_ptr;
+	uint8_t current_channel_id;
+#if CONFIG_ADC_ASYNC
+	struct k_thread thread;
+
+	K_KERNEL_STACK_MEMBER(stack, CONFIG_ADC_MAX11102_17_ACQUISITION_THREAD_STACK_SIZE);
+#endif /* CONFIG_ADC_ASYNC */
+};
+
+static int max11102_17_switch_channel(const struct device *dev)
+{
+	const struct max11102_17_config *config = dev->config;
+	struct max11102_17_data *data = dev->data;
+	int result;
+	uint8_t buffer_rx[1];
+	const struct spi_buf rx_buf[] = {{
+		.buf = buffer_rx,
+		.len = ARRAY_SIZE(buffer_rx),
+	}};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+	struct spi_dt_spec bus;
+
+	memcpy(&bus, &config->bus, sizeof(bus));
+	bus.config.operation |= SPI_HOLD_ON_CS;
+
+	result = spi_read_dt(&bus, &rx);
+	if (result != 0) {
+		LOG_ERR("read failed with error %i", result);
+		return result;
+	}
+
+	gpio_pin_set_dt(&config->gpio_chsel, data->current_channel_id);
+
+	result = spi_read_dt(&config->bus, &rx);
+	if (result != 0) {
+		LOG_ERR("read failed with error %i", result);
+		return result;
+	}
+
+	return 0;
+}
+
+static int max11102_17_channel_setup(const struct device *dev,
+				     const struct adc_channel_cfg *channel_cfg)
+{
+	struct max11102_17_data *data = dev->data;
+	const struct max11102_17_config *config = dev->config;
+
+	LOG_DBG("read from ADC channel %i", channel_cfg->channel_id);
+
+	if (channel_cfg->reference != ADC_REF_EXTERNAL0) {
+		LOG_ERR("invalid reference %i", channel_cfg->reference);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("invalid gain %i", channel_cfg->gain);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("invalid acquisition time %i", channel_cfg->acquisition_time);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential != 0) {
+		LOG_ERR("differential inputs are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id > config->channel_count) {
+		LOG_ERR("invalid channel selection %i", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id != data->current_channel_id) {
+		LOG_DBG("switch channel selection");
+		data->current_channel_id = channel_cfg->channel_id;
+		max11102_17_switch_channel(dev);
```

Comment:
This function should not do channel switching. `adc_channel_setup()` is supposed to only set a configuration for a given channel. Then, the channels to be sampled are specified in the `channels` field of the `adc_sequence` structure pointed in calls to `adc_read()` (or `adc_read_async()`). And users should be able to select multiple channels in one call to `adc_read()` if a given ADC device features multiple channels. So the actual channel switching should be done when a sampling sequence is performed.

Reply:
done